# Hacker News

In [5]:
#2 read in

from csv import reader

opened_file = open('hacker_news.csv')
read_file = reader(opened_file)
hn = list(read_file)
hn_header = hn[0]
hn = hn[1:]


In [12]:
hn_header

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

In [18]:
# hn[0:5]

In [29]:
#3 Extract "Ask HN" and "Show HN" posts

ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    # print(title)
    if title.startswith('Ask HN'):
        ask_posts.append(row)       
    elif title.startswith('Show HN'):
        show_posts.append(row)
    else:
        other_posts.append(row)

In [30]:
print('ask_posts len',len(ask_posts))
print('show_posts len',len(show_posts))
print('other_posts len',len(other_posts))

ask_posts len 1742
show_posts len 1161
other_posts len 17197


In [35]:
# ask_posts[0:5]
# show_posts[0:5]
# other_posts[0:5]

In [68]:
#4 Calculating the average number of comments for Ask HN and Show HN posts

total_ask_comments = 0 
len_ask_comments = 0

for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    len_ask_comments += 1
    
avg_ask_comments = total_ask_comments/len_ask_comments
print('avg_ask_comments',avg_ask_comments)

avg_ask_comments 14.044776119402986


In [58]:
total_show_comments = 0 
len_show_comments = 0

for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments
    len_show_comments += 1

avg_show_comments = total_show_comments/len_show_comments
print('avg_show_comments',avg_show_comments)

avg_show_comments 10.324720068906116


*Ask HN* posts receive more posts on average (14), than *Show HN* posts (10).

In [121]:
#5 Finding the number of ask posts and comments by hour created

import datetime as dt

result_list = []

for row in ask_posts:
    created_at = row[6]
    num_comments = int(row[4])
    result_list.append([created_at, num_comments])
result_list[0:5]

[['8/16/2016 9:55', 6],
 ['11/22/2015 13:43', 29],
 ['5/2/2016 10:14', 1],
 ['8/2/2016 14:20', 3],
 ['10/15/2015 16:38', 17]]

In [110]:
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    hour_str = row[0]  
    hour_dt = dt.datetime.strptime(hour_str, '%m/%d/%Y %H:%M')
    time = hour_dt.hour
    comment = row[1]

    if time not in counts_by_hour:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment
    else:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment

In [117]:
counts_by_hour

{9: 45,
 13: 85,
 10: 59,
 14: 107,
 16: 108,
 23: 68,
 12: 73,
 17: 100,
 15: 116,
 21: 109,
 20: 80,
 2: 58,
 18: 108,
 3: 54,
 5: 46,
 19: 110,
 1: 60,
 22: 71,
 8: 48,
 4: 47,
 0: 54,
 6: 44,
 7: 34,
 11: 58}

`19: 110` The hour with the most posts was 7 pm. 


In [116]:
comments_by_hour

{9: 251,
 13: 1253,
 10: 793,
 14: 1416,
 16: 1814,
 23: 543,
 12: 687,
 17: 1146,
 15: 4477,
 21: 1745,
 20: 1722,
 2: 1381,
 18: 1430,
 3: 421,
 5: 464,
 19: 1188,
 1: 683,
 22: 479,
 8: 492,
 4: 337,
 0: 439,
 6: 397,
 7: 267,
 11: 641}

`15: 4477` The hour with the most comments was 3 pm.